Batch remove

Alexey-ebi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Put closes #XXXX in your comment to auto-close the issue that your -->
<!--- PR fixes (if such).Please link to the issue here: -->
partially address #16 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change which improve code itself)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Alexey-ebi committed:
Starting to work on batch remove

Alexey-ebi committed:
Adding tasks for batch update for animals and samples

Alexey-ebi committed:
Refactoring tasks

Alexey-ebi committed:
Adding logic to tasks

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,107 @@
+    # Ovverride default on failure method
+    # This is not a failed validation for a wrong value, this is an
+    # error in task that mean an error in coding
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
```
This seems to be used many time and in my opinion it will become part of a TaskFailureMixin. For the moment, it's ok like this

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,100 @@
+        failed_ids = list()
+        for sample_id in sample_ids:
+            try:
+                name = Name.objects.get(name=sample_id)
```
You need to consider submission when selecting Name, or you will get an error like `Error in batch delete for samples: get() returned more than one Name -- it returned X!` when deleting samples with the same name (as when uploading submission or part of submission in your profile or when you upload the same dataset of another one - ie: the italian subset. Try to modify like this:
```python
submission = Submission.objects.get(pk=submission_id)
name = Name.objects.get(name=sample_id, submission=submission)
```
The rest of the function could work as it its. You could test this code by uploading two submissions and then deleting from one, ensuring the second remain the same

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,100 @@
+        for sample_id in sample_ids:
+            try:
+                name = Name.objects.get(name=sample_id)
+                object = Sample.objects.get(name=name)
```
`object` is a reserved key in python (if the base of each class!). If you see`self.object` in my code is because every `DetailView` as an attribute related to its model. Please, give it a different name

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -284,6 +287,54 @@ def form_valid(self, form):
+        summary_obj.reset()
+        res = my_task.delay(pk, [item for item in keys_to_delete])
+        logger.info(
+            "Start fix validation process with task %s" % res.task_id)
```
Please change this message

bunop commented on django-data/image/submissions/templates/submissions/submission_batch_delete.html:
```diff
@@ -0,0 +1,17 @@
+    <form action="{% url 'submissions:batch_delete' pk delete_type%}" method="post">
+        {% csrf_token %}
+        <div class="form-group">
+            <label for="to_delete">Please provide ids to delete (one id per row)</label>
```
If you use the `forms` django methods, fields, ids, and rendering of forms itself is managed by templates and forms widgets

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,107 @@
+        failed_ids = list()
+        for animal_id in animal_ids:
+            try:
+                name = Name.objects.get(name=animal_id)
```
You could have the same animal names in different submission. Check my comments for `samples.tasks`

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -284,6 +287,54 @@ def form_valid(self, form):
+                                               'pk': pk})
+
+
+class BatchDelete(OwnerMixin, View):
```
Instead of working with the base `View` class, you could use the `FormView` class. The `get` and `post` low levels methods are defined by default accordingly to `forms.Form` class; this `post` method could change in `form_valid` in which you will do all this stuff. `DeleteAnimalsView` and `DeleteSamplesView`, which render only the empty form, could be removed and the `delete_type` of removal could be managed by url and templates. The redirect you return, will become the `success_url` attribute class

bunop replied:
This is only a suggestion on how to render code more `django`style

bunop commented on django-data/image/submissions/urls.py:
```diff
@@ -33,7 +33,16 @@
     url(r'^(?P<pk>[-\w]+)/delete/$',
         views.DeleteSubmissionView.as_view(),
         name='delete'),
+    url(r'^(?P<pk>[-\w]+)/delete_animals/$',
```
Those two links could be removed using the django `FormView`. See my comments on `submissions.views`

bunop reviewed (changes requested):
Ty @Alexey-ebi for this PR. There's one thing that must be changed: you must ensure that the Name you fecth during batch removal belong to the submission you are trying to remove, or otherwise you can't get an unique object. Other comments are not critical but could be more django-style. If you don't have time to fix this, let me know; I will complete and accepts those pull request in these days or when I will return from holidays. Ty again

bunop committed:
:twisted_rightwards_arrows: Merge branch 'master' into batch_remove

bunop committed:
:bug: fix and test sample tasks

bunop committed:
:bug: fix and test animal tasks

bunop committed:
:white_check_mark: testing issues with sample/animal batch delete

bunop committed:
:bug: fixing calling exceptions in animals/samples batch delete

bunop committed:
:recycle: move submission delete tests mixin in common module

bunop committed:
:white_check_mark: test submission ownership for batch delete

The base View class doesn't have the get_queryset method, so is
unable to verify the ownership of the submission that I need to
modify with DeleteAnimalsView and DeleteSamplesView

bunop committed:
:white_check_mark: testing BatchDelete view
